src/node_buffer.cc

```diff
@@ -994,10 +996,11 @@ void IndexOfString(const FunctionCallbackInfo<Value>& args) {
   bool is_forward = args[4]->IsTrue();
 
   const char* haystack = ts_obj_data;
-  const size_t haystack_length = ts_obj_length;
-  // Extended latin-1 characters are 2 bytes in Utf8.
-  const size_t needle_length =
-      enc == BINARY ? needle->Length() : needle->Utf8Length();
+  // Round down to the neares multiple of 2 in case of UCS2.
+  const size_t haystack_length = enc == UCS2 ?
+    ts_obj_length & ~1 : ts_obj_length;
+
+  const size_t needle_length = StringBytes::Size(env->isolate(), needle, enc);
```

Comment:
As @bnoordhuis helpfully reminded me in a different PR recently, you can use `args.GetIsolate()` here and avoid having to add the `Environment* env = Environment::GetCurrent(args)` above on line 985. See line 987 also :-).